cms/djangoapps/contentstore/outlines.py

```diff
@@ -0,0 +1,116 @@
+"""
+This is where Studio interacts with the learning_sequences application, which
+is responsible for holding course outline data. Studio _pushes_ that data into
+learning_sequences at publish time.
+"""
+from datetime import timezone
+
+from edx_django_utils.monitoring import function_trace, set_custom_attribute
+
+from openedx.core.djangoapps.content.learning_sequences.api import replace_course_outline
+from openedx.core.djangoapps.content.learning_sequences.data import (
+    CourseLearningSequenceData,
+    CourseOutlineData,
+    CourseSectionData,
+    CourseVisibility,
+    ExamData,
+    VisibilityData,
+)
+from xmodule.modulestore import COURSE_ROOT, LIBRARY_ROOT, ModuleStoreEnum
+from xmodule.modulestore.django import modulestore
+
+
+@function_trace('get_outline_from_modulestore')
+def get_outline_from_modulestore(course_key):
+    """
+    Get a learning_sequence.data.CourseOutlineData for a param:course_key
+    """
+    def _remove_version_info(usage_key):
+        """
+        When we ask modulestore for the published branch in the Studio process
+        after catching a publish signal, the items that have been changed will
+        return UsageKeys that have full version information in their attached
+        CourseKeys. This makes them hash and serialize differently. We want to
+        strip this information and have everything use a CourseKey with no
+        version information attached.
+
+        IF we can verify that this property is still true when running in an
+        out-of-process celery task, it might be a path towards optimizing our
+        publish updates to only what has changed–but that carries other risks as
+        well in terms of correctness (e.g. race conditions).
+        """
+        return usage_key.map_into_course(course_key)
+
+    def _make_section_data(section):
+        sequences_data = []
+        for sequence in section.get_children():
+            sequences_data.append(
+                CourseLearningSequenceData(
+                    usage_key=_remove_version_info(sequence.location),
+                    title=sequence.display_name,
+                    inaccessible_after_due=sequence.hide_after_due,
+                    exam=ExamData(
+                        is_practice_exam=sequence.is_practice_exam,
+                        is_proctored_enabled=sequence.is_proctored_enabled,
+                        is_time_limited=sequence.is_time_limited,
+                    ),
+                    visibility=VisibilityData(
+                        hide_from_toc=sequence.hide_from_toc,
+                        visible_to_staff_only=sequence.visible_to_staff_only,
```

Comment:
[suggestion] I think this function assumes that the course will have zero or more ChapterBlocks as children, each of which will have zero or more SequentialBlocks as children. I know that those are reasonable assumptions for any edx.org course, but do you think it'd be worth checking anyway?

On the off chance that edx.org or some other instance had a weirdly-structured course imported from OLX, it would be nice if the error were something like `"section abc in course xyz has unexpected child of type <blah>; will not generate outline"` instead of `"attributeerror: is_practice_exam"`.

Reply:
Good point. I'll make it a bit more resilient.